File: apps/web/src/api/agent.ts
import { apiRequest } from "./http";
import type { ResourceMutationContext } from "./resources";

export type AgentScopeDomain = "goals" | "tasks" | "calendar" | "records" | "notes";
export type ServerAgentRunStatus = "ready" | "reading" | "analyzing" | "waiting" | "applying" | "completed" | "failed" | "stopped";
export type ServerAgentStepStatus = "pending" | "running" | "done" | "failed";
export type ServerAgentChangeStatus = "pending" | "accepted" | "rejected" | "applied" | "failed" | "conflicted";

export interface AgentScope {
  domains: AgentScopeDomain[];
  entityIds?: string[];
  from?: string;
  to?: string;
}

export interface ServerAgentStep {
  id: string;
  runId: string;
  sequenceNo: number;
  title: string;
  detail: string;
  status: ServerAgentStepStatus;
  metadata: Record<string, unknown>;
  startedAt?: string;
  finishedAt?: string;
  version: number;
  createdAt: string;
  updatedAt: string;
}

export interface ServerAgentChange {
  id: string;
  runId: string;
  changeType: string;
  targetType: string;
  targetId?: string;
  baseVersion?: number;
  patch: unknown[];
  previewBefore?: unknown;
  previewAfter?: unknown;
  reason: string;
  status: ServerAgentChangeStatus;
  acceptedAt?: string;
  appliedAt?: string;
  version: number;
  createdAt: string;
  updatedAt: string;
}

export interface ServerAgentSourceRef {
  id: string;
  runId: string;
  entityType: string;
  entityId: string;
  entityVersion: number;
  labelSnapshot: string;
  createdAt: string;
}

export interface ServerAgentRun {
  id: string;
  intent: string;
  status: ServerAgentRunStatus;
  actionMode: "read" | "confirm";
  scope: AgentScope;
  provider?: string;
  model?: string;
  startedAt?: string;
  finishedAt?: string;
  summary?: string;
  errorCode?: string;
  errorMessage?: string;
  version: number;
  createdAt: string;
  updatedAt: string;
  steps: ServerAgentStep[];
  changes: ServerAgentChange[];
  sourceRefs: ServerAgentSourceRef[];
}

export interface AgentRunPage {
  runs: ServerAgentRun[];
  nextCursor?: string;
  hasMore: boolean;
}

export interface CreateAgentRunInput {
  intent: string;
  actionMode: "read" | "confirm";
  scope: AgentScope;
}

export interface AgentChangeResult {
  change: ServerAgentChange;
  run: ServerAgentRun;
}

function mutationHeaders(context: ResourceMutationContext, version?: number): HeadersInit {
  return {
    "X-Device-ID": context.deviceId,
    "Idempotency-Key": context.mutationId,
    ...(version === undefined ? {} : { "If-Match": `"${version}"` }),
  };
}

export function listAgentRuns(options: { cursor?: string; limit?: number } = {}): Promise<AgentRunPage> {
  const query = new URLSearchParams();
  if (options.cursor) query.set("cursor", options.cursor);
  if (options.limit !== undefined) query.set("limit", String(options.limit));
  const encoded = query.toString();
  return apiRequest(`/agent-runs${encoded ? `?${encoded}` : ""}`);
}

export function createAgentRun(input: CreateAgentRunInput, context: ResourceMutationContext): Promise<ServerAgentRun> {
  return apiRequest("/agent-runs", { method: "POST", headers: mutationHeaders(context), json: input });
}

export function getAgentRun(runId: string): Promise<ServerAgentRun> {
  return apiRequest(`/agent-runs/${runId}`);
}

export function stopAgentRun(runId: string, version: number, context: ResourceMutationContext): Promise<ServerAgentRun> {
  return apiRequest(`/agent-runs/${runId}/stop`, { method: "POST", headers: mutationHeaders(context, version) });
}

export function acceptAgentChange(changeId: string, version: number, context: ResourceMutationContext): Promise<AgentChangeResult> {
  return apiRequest(`/agent-changes/${changeId}/accept`, { method: "POST", headers: mutationHeaders(context, version) });
}

export function rejectAgentChange(changeId: string, version: number, context: ResourceMutationContext): Promise<AgentChangeResult> {
  return apiRequest(`/agent-changes/${changeId}/reject`, { method: "POST", headers: mutationHeaders(context, version) });
}

File: apps/api/internal/model/agent.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type AgentScope struct {
	Domains   []string    `json:"domains"`
	EntityIDs []uuid.UUID `json:"entityIds,omitempty"`
	From      *time.Time  `json:"from,omitempty"`
	To        *time.Time  `json:"to,omitempty"`
}

type AgentRun struct {
	ID           uuid.UUID        `json:"id"`
	Intent       string           `json:"intent"`
	Status       string           `json:"status"`
	ActionMode   string           `json:"actionMode"`
	Scope        json.RawMessage  `json:"scope"`
	Provider     *string          `json:"provider,omitempty"`
	Model        *string          `json:"model,omitempty"`
	StartedAt    *time.Time       `json:"startedAt,omitempty"`
	FinishedAt   *time.Time       `json:"finishedAt,omitempty"`
	Summary      *string          `json:"summary,omitempty"`
	ErrorCode    *string          `json:"errorCode,omitempty"`
	ErrorMessage *string          `json:"errorMessage,omitempty"`
	Version      int64            `json:"version"`
	CreatedAt    time.Time        `json:"createdAt"`
	UpdatedAt    time.Time        `json:"updatedAt"`
	Steps        []AgentStep      `json:"steps"`
	Changes      []AgentChange    `json:"changes"`
	SourceRefs   []AgentSourceRef `json:"sourceRefs"`
}

type AgentStep struct {
	ID         uuid.UUID       `json:"id"`
	RunID      uuid.UUID       `json:"runId"`
	SequenceNo int             `json:"sequenceNo"`
	Title      string          `json:"title"`
	Detail     string          `json:"detail"`
	Status     string          `json:"status"`
	Metadata   json.RawMessage `json:"metadata"`
	StartedAt  *time.Time      `json:"startedAt,omitempty"`
	FinishedAt *time.Time      `json:"finishedAt,omitempty"`
	Version    int64           `json:"version"`
	CreatedAt  time.Time       `json:"createdAt"`
	UpdatedAt  time.Time       `json:"updatedAt"`
}

type AgentChange struct {
	ID            uuid.UUID       `json:"id"`
	RunID         uuid.UUID       `json:"runId"`
	ChangeType    string          `json:"changeType"`
	TargetType    string          `json:"targetType"`
	TargetID      *uuid.UUID      `json:"targetId,omitempty"`
	BaseVersion   *int64          `json:"baseVersion,omitempty"`
	Patch         json.RawMessage `json:"patch"`
	PreviewBefore json.RawMessage `json:"previewBefore,omitempty"`
	PreviewAfter  json.RawMessage `json:"previewAfter,omitempty"`
	Reason        string          `json:"reason"`
	Status        string          `json:"status"`
	AcceptedAt    *time.Time      `json:"acceptedAt,omitempty"`
	AppliedAt     *time.Time      `json:"appliedAt,omitempty"`
	Version       int64           `json:"version"`
	CreatedAt     time.Time       `json:"createdAt"`
	UpdatedAt     time.Time       `json:"updatedAt"`
}

type AgentSourceRef struct {
	ID            uuid.UUID `json:"id"`
	RunID         uuid.UUID `json:"runId"`
	EntityType    string    `json:"entityType"`
	EntityID      uuid.UUID `json:"entityId"`
	EntityVersion int64     `json:"entityVersion"`
	LabelSnapshot string    `json:"labelSnapshot"`
	CreatedAt     time.Time `json:"createdAt"`
}

type AgentStepDraft struct {
	Title    string          `json:"title"`
	Detail   string          `json:"detail"`
	Metadata json.RawMessage `json:"metadata"`
}

type AgentChangeDraft struct {
	ChangeType    string          `json:"changeType"`
	TargetType    string          `json:"targetType"`
	TargetID      *uuid.UUID      `json:"targetId,omitempty"`
	BaseVersion   *int64          `json:"baseVersion,omitempty"`
	Patch         json.RawMessage `json:"patch"`
	PreviewBefore json.RawMessage `json:"previewBefore,omitempty"`
	PreviewAfter  json.RawMessage `json:"previewAfter,omitempty"`
	Reason        string          `json:"reason"`
}

type AgentSourceRefDraft struct {
	EntityType    string    `json:"entityType"`
	EntityID      uuid.UUID `json:"entityId"`
	EntityVersion int64     `json:"entityVersion"`
	LabelSnapshot string    `json:"labelSnapshot"`
}

type AgentPlan struct {
	Summary    string                `json:"summary"`
	Steps      []AgentStepDraft      `json:"steps"`
	Changes    []AgentChangeDraft    `json:"changes"`
	SourceRefs []AgentSourceRefDraft `json:"sourceRefs"`
}

type AgentSnapshot struct {
	Run     AgentRun
	Goals   []Goal
	Tasks   []Task
	Events  []CalendarEvent
	Records []Record
	Notes   []Note
}

type AgentApplyResult struct {
	Change          AgentChange     `json:"change"`
	Run             AgentRun        `json:"run"`
	RunUpdated      bool            `json:"-"`
	TargetType      string          `json:"-"`
	TargetID        uuid.UUID       `json:"-"`
	TargetOperation string          `json:"-"`
	TargetVersion   int64           `json:"-"`
	BeforeData      json.RawMessage `json:"-"`
	AfterData       json.RawMessage `json:"-"`
}
